pkg/controllers/management/auth/globalroles/globalrole_handler.go

```diff
@@ -196,6 +214,95 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
 	return nil
 }
 
+func (gr *globalRoleLifecycle) reconcileNamespacedRoles(globalRole *v3.GlobalRole) error {
+	var returnError error
+
+	// get all the roles claiming to be owned by this GR and remove any that shouldn't exist
+	r, _ := labels.NewRequirement(grOwnerLabel, selection.Equals, []string{globalRole.Name})
+	roles, err := gr.rLister.List("", labels.NewSelector().Add(*r))
+	if err != nil {
+		returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't list roles with label %s : %s", grOwnerLabel, globalRole.Name))
+	} else {
+		err = gr.purgeInvalidRoles(roles, globalRole.NamespacedRules)
+		if err != nil {
+			returnError = multierror.Append(returnError, err)
+		}
+	}
+
+	for ns, rules := range globalRole.NamespacedRules {
+		_, err = gr.nsLister.Get("", ns)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
+		// Check if the role exists
+		roleName := globalRole.Name + "-" + ns
+		role, err := gr.rLister.Get(ns, roleName)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+			newRole := &v1.Role{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      roleName,
+					Namespace: ns,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: globalRole.APIVersion,
+							Kind:       globalRole.Kind,
+							Name:       globalRole.Name,
+							UID:        globalRole.UID,
+						},
+					},
+					Labels: map[string]string{
+						grOwnerLabel: globalRole.Name,
+					},
+				},
+				Rules: rules,
+			}
+			_, err = gr.rClient.Create(newRole)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
I think we need to validate the integrity of the existing role if it already exists. I would recommend removing the else below and allowing the already exist case to fall through to the below logic.